I2s loopback config

Raymond0225 opened the pull request:
Use TX and RX blocks on SAI for i2s speed test, by this way we can avoid HW reworking.
    Update Readme file and DMA channel configuration in overlay file to avoid DMA channel
    conflict with other peripherals like UART.
  

    In previous implementions, when there is no buffers in TX queue in a TX
    callback context, TX will be stopped immediately. It is not correct
    because there may still TX blocks in DMA TX transfer queue.
    TX should only be stopped when DMA queue is empty and no more in I2S TX
    queue.
    Another modification is to set correct TX water FIFO level.

   require this HAL patch to pass the test:
   https://github.com/zephyrproject-rtos/hal_nxp/pull/490

dleach02 commented:
@Raymond0225 add an update to this PR the west.yml to point to the needed HAL PR.

Raymond0225 commented:
> @Raymond0225 add an update to this PR the west.yml to point to the needed HAL PR.

To my understanding, we need merge the HAL PR firstly then update the full SHA in west.yml, right?
Will the SHA in the PR be changed after merge?

Raymond0225 commented:
Resolve conflict with last main source code, rebase to latest code of upstream.

Raymond0225 committed:
drivers: i2s: Fix TX transfer stopping issue in some corner cases

In previous implementions, when there is no buffers in TX queue in a TX
callback context, TX will be stopped immediately. It is not correct
because there may still TX blocks in DMA TX transfer queue.
TX should only be stopped when DMA queue is empty and no more in I2S TX
queue.
Another modification is to set correct TX water FIFO level.

Signed-off-by: Raymond Lei <raymond.lei@nxp.com>

Raymond0225 committed:
test: i2s: Reconfigure i2s speed test HW environment on RT1170 EVK board

Use TX and RX blocks on SAI for i2s speed test, by this way we can avoid
HW reworking.
Update Readme file and DMA channel configuration in overlay file to avoid
DMA channel conflict with other peripherals like UART.

Signed-off-by: Raymond Lei <raymond.lei@nxp.com>

decsny commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -308,11 +308,23 @@ static void i2s_dma_tx_callback(const struct device *dma_dev, void *arg, uint32_
 		strm->state = I2S_STATE_READY;
 		LOG_DBG("TX stream has stopped");
+		goto disabled_exit_no_drop;
 	} else {
```
I think the else might not be necessary, should we remove it to reduce nesting ?

Raymond0225 replied:
yes, we can remove it.

mmahadevan108 commented on tests/drivers/i2s/i2s_speed/boards/mimxrt1170_evk_mimxrt1176_cm7.overlay:
```diff
@@ -1,26 +1,15 @@
- */
-&enet {
-	status = "disabled";
+/* Default DMA channel 0 and 1 maybe used by other peripherals like UART. */
```
We may not need this as this test is probably run standalone.

Raymond0225 replied:
I know it is running alone but when logging is enabled, UART1 works by default and this will impact DMA channel 0/1.
I tried, test will fail if we do not override DMA channels of UART1.

mmahadevan108 commented on tests/drivers/i2s/i2s_speed/boards/mimxrt1170_evk_mimxrt1176_cm7_B.overlay:
```diff
@@ -0,0 +1,15 @@
+	};
+};
+
+/* Default DMA channel 0 and 1 maybe used by other peripherals like UART. */
```
Same comment as above

mmahadevan108 commented:
@anangl , can you help review.
